zkevm-circuits/src/table.rs

```diff
@@ -0,0 +1,874 @@
+//! Table definitions used cross-circuits
+
+use crate::copy_circuit::number_or_hash_to_field;
+use crate::evm_circuit::witness::RwRow;
+use crate::evm_circuit::{
+    util::{rlc, RandomLinearCombination},
+    witness::{Block, BlockContext, Bytecode, RwMap, Transaction},
+};
+use crate::impl_expr;
+use bus_mapping::circuit_input_builder::{CopyDataType, CopyEvent};
+use eth_types::{Field, ToAddress, ToLittleEndian, ToScalar, Word, U256};
+use gadgets::binary_number::{BinaryNumberChip, BinaryNumberConfig};
+use halo2_proofs::{
+    arithmetic::FieldExt,
+    circuit::Region,
+    plonk::{Advice, Column, ConstraintSystem, Error},
+};
+use halo2_proofs::{circuit::Layouter, plonk::*, poly::Rotation};
+use itertools::Itertools;
+use keccak256::plain::Keccak;
+use strum_macros::{EnumCount, EnumIter};
+
+/// Trait used for dynamic tables.  Used to get an automatic implementation of
+/// the LookupTable trait where each `table_expr` is a query to each column at
+/// `Rotation::cur`.
+pub trait TableColumns {
+    /// Returns the list of advice columns following the table order.
+    fn columns(&self) -> Vec<Column<Advice>>;
+}
```

Comment:
Perhaps we can give it a more specific name like `DynamicTableColumns`? Otherwise it looks similar to builtin `TableColumn`.

Reply:
Thanks for the suggestion! Updated in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/622/commits/7955480c55c5be39fc7efb1710ca2dbf58e34ec8